src/Internal/Payment/State/PaymentState.php

```diff
@@ -0,0 +1,117 @@
+<?php
+/**
+ * Class PaymentState
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay\Internal\Payment\State;
+
+use WCPay\Internal\Payment\Exception\MethodUnavailableException;
+use WCPay\Internal\Payment\Exception\StateTransitionException;
+use WCPay\Internal\Payment\PaymentContext;
+
+/**
+ * Base class for payment states.
+ */
+abstract class PaymentState {
+	/**
+	 * Holds the context of the payments.
+	 *
+	 * The context contains working payment data,
+	 * while states are purely functional and stateless.
+	 *
+	 * @var PaymentContext
+	 */
+	private $context;
+
+	/**
+	 * A factory for states.
+	 *
+	 * States can
+	 *
+	 * @var StateFactory
+	 */
+	private $state_factory;
+
+	/**
+	 * Class constructor, only meant for storing dependencies.
+	 *
+	 * @param StateFactory $state_factory Factory for payment states.
+	 */
+	public function __construct( StateFactory $state_factory ) {
+		$this->state_factory = $state_factory;
+	}
+
+	/**
+	 * Stores the payment context.
+	 *
+	 * @param PaymentContext $context Payment context.
+	 */
+	public function set_context( PaymentContext $context ) {
+		$this->context = $context;
+	}
+
+	/**
+	 * Returns the payment context.
+	 *
+	 * @return PaymentContext
+	 */
+	public function get_context(): PaymentContext {
+		return $this->context;
+	}
+
+	/**
+	 * Creates a new instance of a given payment state class.
+	 *
+	 * States control the payment flow, and allow transitions to the next state.
+	 * This method should only be called whenever the process is ready to transition
+	 * to the next state, as each new state will be considered the payment's latest one.
+	 *
+	 * @param string $state_class The class of the state to crate.
+	 * @return PaymentState
+	 * @throws StateTransitionException In case the new state could not be created.
+	 */
+	protected function create_state( string $state_class ) {
+		$state = $this->state_factory->create_state( $state_class, $this->context );
```

Comment:
IMHO, the state should not set context. Context should be a godlike class that shifts states and keeps all data in between. However, feel free to disregard this comment if you prefer a current approach.

Reply:
You are continuing the "state pattern vs. state machine" discussion we were having at the meetup.  At the end, we agreed to try the new approach and avoid having `Payment` as the god-like class, working with states directly.

So far, that seems like it's working, but I'm curious to see whether something in this PR or the follow-up ones seems like a bad idea to you.